Update accumulate_grad to support vmap

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49120 Add batched grad testing to gradcheck, turn it on in test_autograd
* **#49119 Update accumulate_grad to support vmap**
* #49118 Move inplace_is_vmap_compatible to BatchedTensorImpl.h

I don't know how the accumulate_grad code gets hit via calling
autograd.grad, so I went through all places in accumulate_grad
that are definitely impossible to vmap through and changed them.

To support this:
- I added vmap support for Tensor::strides(). It returns the strides
that correspond to the public dimensions of the tensor (not the ones
being vmapped over).
- Changed an instance of empty_strided to new_empty_strided.
- Replaced an in-place operation in accumulate_grad.h

Test Plan:
- added a test for calling strides() inside of vmap
- added tests that exercise all of the accumulate_grad code path.
NB: I don't know why these tests exercise the code paths, but I've
verified that they do via gdb.

Suggestions for some saner test cases are very welcome.

Differential Revision: [D25563543](https://our.internmc.facebook.com/intern/diff/D25563543)

zou3519 committed:
Update accumulate_grad to support vmap

I don't know how the accumulate_grad code gets hit, so I went through
all places in accumulate_grad that are definitely impossible to vmap
through and changed them.

To support this:
- I added vmap support for Tensor::strides(). It returns the strides
that correspond to the public dimensions of the tensor (not the ones
being vmapped over).
- Changed an instance of empty_strided to new_empty_strided.
- Replaced an in-place operation in accumulate_grad.h

Test Plan:
- added a test for calling strides() inside of vmap
- added a test that exercise some of the accumulate grad code path. I
don't know if it exercises all of the code changes in this PR.
Suggestions for more test cases are very welcome.

[ghstack-poisoned]

zou3519 committed:
Update on "Update accumulate_grad to support vmap"


I don't know how the accumulate_grad code gets hit, so I went through
all places in accumulate_grad that are definitely impossible to vmap
through and changed them.

To support this:
- I added vmap support for Tensor::strides(). It returns the strides
that correspond to the public dimensions of the tensor (not the ones
being vmapped over).
- Changed an instance of empty_strided to new_empty_strided.
- Replaced an in-place operation in accumulate_grad.h

Test Plan:
- added a test for calling strides() inside of vmap
- added a test that exercise some of the accumulate grad code path. I
don't know if it exercises all of the code changes in this PR.
Suggestions for more test cases are very welcome.

[ghstack-poisoned]

zou3519 committed:
Update on "Update accumulate_grad to support vmap"


I don't know how the accumulate_grad code gets hit, so I went through
all places in accumulate_grad that are definitely impossible to vmap
through and changed them.

To support this:
- I added vmap support for Tensor::strides(). It returns the strides
that correspond to the public dimensions of the tensor (not the ones
being vmapped over).
- Changed an instance of empty_strided to new_empty_strided.
- Replaced an in-place operation in accumulate_grad.h

Test Plan:
- added a test for calling strides() inside of vmap
- added a test that exercise some of the accumulate grad code path. I
don't know if it exercises all of the code changes in this PR.
Suggestions for more test cases are very welcome.

[ghstack-poisoned]

zou3519 committed:
Update on "Update accumulate_grad to support vmap"


I don't know how the accumulate_grad code gets hit, so I went through
all places in accumulate_grad that are definitely impossible to vmap
through and changed them.

To support this:
- I added vmap support for Tensor::strides(). It returns the strides
that correspond to the public dimensions of the tensor (not the ones
being vmapped over).
- Changed an instance of empty_strided to new_empty_strided.
- Replaced an in-place operation in accumulate_grad.h

Test Plan:
- added a test for calling strides() inside of vmap
- added tests that exercise all of the accumulate_grad code path.
NB: I don't know why these tests exercise the code paths, but I've
verified that they do via gdb.

Suggestions for some saner test cases are very welcome.

[ghstack-poisoned]

zou3519 committed:
Update on "Update accumulate_grad to support vmap"


I don't know how the accumulate_grad code gets hit, so I went through
all places in accumulate_grad that are definitely impossible to vmap
through and changed them.

To support this:
- I added vmap support for Tensor::strides(). It returns the strides
that correspond to the public dimensions of the tensor (not the ones
being vmapped over).
- Changed an instance of empty_strided to new_empty_strided.
- Replaced an in-place operation in accumulate_grad.h

Test Plan:
- added a test for calling strides() inside of vmap
- added tests that exercise all of the accumulate_grad code path.
NB: I don't know why these tests exercise the code paths, but I've
verified that they do via gdb.

Suggestions for some saner test cases are very welcome.

[ghstack-poisoned]

zou3519 committed:
Update on "Update accumulate_grad to support vmap"


I don't know how the accumulate_grad code gets hit via calling
autograd.grad, so I went through all places in accumulate_grad
that are definitely impossible to vmap through and changed them.

To support this:
- I added vmap support for Tensor::strides(). It returns the strides
that correspond to the public dimensions of the tensor (not the ones
being vmapped over).
- Changed an instance of empty_strided to new_empty_strided.
- Replaced an in-place operation in accumulate_grad.h

Test Plan:
- added a test for calling strides() inside of vmap
- added tests that exercise all of the accumulate_grad code path.
NB: I don't know why these tests exercise the code paths, but I've
verified that they do via gdb.

Suggestions for some saner test cases are very welcome.

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchedTensorImpl.cpp:
```diff
@@ -76,13 +76,6 @@ void BatchedTensorImpl::checkInvariants() const {
-}
-int64_t BatchedTensorImpl::stride(int64_t d) const {
-  TORCH_CHECK(false, "NYI: Getting tensor strides inside of vmap");
-}
```
cc @ngimel  YASSSSSSS

ezyang commented:
So, now that vmap supports strides, there was something I wasn't too clear about. Suppose that I have a contiguous tensor with strides (8, 4, 2, 1) and let's say I vmap with batch dim = 2 (e.g., not the first dimension). Is the BatchedTensor contiguous? (I don't think it should be!) I couldn't tell from this PR what would happen in this case.

zou3519 commented:
> So, now that vmap supports strides, there was something I wasn't too clear about. Suppose that I have a contiguous tensor with strides (8, 4, 2, 1) and let's say I vmap with batch dim = 2 (e.g., not the first dimension). Is the BatchedTensor contiguous? (I don't think it should be!) I couldn't tell from this PR what would happen in this case.

It's not contiguous. The reason behind this is asking "is a BatchedTensor contiguous" is equivalent to asking "are the per-examples of the BatchedTensor contiguous". The per-examples in your example would have strides (8, 4, 1) which make them not contiguous

ngimel commented on torch/csrc/autograd/utils/grad_layout_contract.h:
```diff
@@ -25,7 +25,7 @@ inline at::Tensor clone_obey_contract(const at::Tensor& new_grad, const at::Tens
     // Does this dicey-looking sequence attach the result to new_grad's
     // history if GradMode::is_enabled()?  Yes, and @alband says it should.
-    return std::move(at::empty_strided(variable.sizes(), variable.strides(),
+    return std::move(new_grad.new_empty_strided(variable.sizes(), variable.strides(),
```
is it because `empty_strided` cannot create BatchedTensor?

zou3519 replied:
Yes, empty_strided doesn't know how to create a BatchedTensor in this situation because it doesn't accept a tensor as an argument. To make up for it, our solution is to use `new_empty_strided`: new_empty_strided can accept BatchedTensors and if it is passed one, it returns a BatchedTensor.

zou3519 committed:
Update on "Update accumulate_grad to support vmap"


I don't know how the accumulate_grad code gets hit via calling
autograd.grad, so I went through all places in accumulate_grad
that are definitely impossible to vmap through and changed them.

To support this:
- I added vmap support for Tensor::strides(). It returns the strides
that correspond to the public dimensions of the tensor (not the ones
being vmapped over).
- Changed an instance of empty_strided to new_empty_strided.
- Replaced an in-place operation in accumulate_grad.h

Test Plan:
- added a test for calling strides() inside of vmap
- added tests that exercise all of the accumulate_grad code path.
NB: I don't know why these tests exercise the code paths, but I've
verified that they do via gdb.

Suggestions for some saner test cases are very welcome.

[ghstack-poisoned]
